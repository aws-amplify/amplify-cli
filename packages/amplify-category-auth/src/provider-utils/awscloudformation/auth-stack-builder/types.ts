export type OAuthMetaData = {
  AllowedOAuthFlows?: Array<string>;
  AllowedOAuthFlowsUserPoolClient?: boolean;
  AllowedOAuthScopes?: Array<string>;
  CallbackURLs?: Array<string>;
  LogoutURLs?: Array<string>;
  SupportedIdentityProviders?: Array<string>;
};
