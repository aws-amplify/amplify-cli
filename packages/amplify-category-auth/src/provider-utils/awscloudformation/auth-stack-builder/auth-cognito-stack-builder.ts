import * as cdk from '@aws-cdk/core';
import * as s3 from '@aws-cdk/aws-s3';
import * as iam from '@aws-cdk/aws-iam';
import { AmplifyAuthCognitoStackTemplate } from './types';
import { CfnUserPool, CfnUserPoolClient, CfnIdentityPool, CfnIdentityPoolRoleAttachment } from '@aws-cdk/aws-cognito';

const CFN_TEMPLATE_FORMAT_VERSION = '2010-09-09';
const ROOT_CFN_DESCRIPTION = 'Root Stack for AWS Amplify CLI';

export type AmplifyAuthCognitoStackProps = {
  synthesizer: cdk.IStackSynthesizer;
};

export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCognitoStackTemplate {
  _scope: cdk.Construct;
  private _cfnParameterMap: Map<string, cdk.CfnParameter> = new Map();
  customMessageConfirmationBucket: s3.CfnBucket | undefined;
  snsRole: iam.CfnRole | undefined;
  userPool: CfnUserPool | undefined;
  userPoolClientWeb: CfnUserPoolClient | undefined;
  userPoolClient: CfnUserPoolClient | undefined;
  userPoolClientRole: iam.CfnRole | undefined;
  identityPool: CfnIdentityPool | undefined;
  identityPoolRoleMap: CfnIdentityPoolRoleAttachment | undefined;

  constructor(scope: cdk.Construct, id: string, props: AmplifyAuthCognitoStackProps) {
    super(scope, id, props);
    this._scope = scope;
    this.templateOptions.templateFormatVersion = CFN_TEMPLATE_FORMAT_VERSION;
    this.templateOptions.description = ROOT_CFN_DESCRIPTION;
  }

  /**
   *
   * @param props :cdk.CfnOutputProps
   * @param logicalId: : lodicalId of the Resource
   */
  addCfnOutput(props: cdk.CfnOutputProps, logicalId: string): void {
    try {
      new cdk.CfnOutput(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnMapping(props: cdk.CfnMappingProps, logicalId: string): void {
    try {
      new cdk.CfnMapping(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnCondition(props: cdk.CfnConditionProps, logicalId: string): void {
    try {
      new cdk.CfnCondition(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }
  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnResource(props: cdk.CfnResourceProps, logicalId: string): void {
    try {
      new cdk.CfnResource(this, logicalId, props);
    } catch (error) {
      throw new Error(error);
    }
  }

  /**
   *
   * @param props
   * @param logicalId
   */
  addCfnParameter(props: cdk.CfnParameterProps, logicalId: string): void {
    try {
      if (this._cfnParameterMap.has(logicalId)) {
        throw new Error('logical Id already Exists');
      }
      this._cfnParameterMap.set(logicalId, new cdk.CfnParameter(this, logicalId, props));
    } catch (error) {
      throw new Error(error);
    }
  }

  getCfnParameter(logicalId: string): cdk.CfnParameter | undefined {
    if (this._cfnParameterMap.has(logicalId)) {
      return this._cfnParameterMap.get(logicalId);
    } else {
      throw new Error(`Cfn Parameter with LogicalId ${logicalId} doesnt exist`);
    }
  }

  generateRootStackResources = async () => {
    // generate Template here
  };

  // add Function for Custom Resource in Root stack
  /**
   *
   * @param _
   * @returns
   */
  public renderCloudFormationTemplate = (_: cdk.ISynthesisSession): string => {
    return JSON.stringify(this._toCloudFormation(), undefined, 2);
  };
}
