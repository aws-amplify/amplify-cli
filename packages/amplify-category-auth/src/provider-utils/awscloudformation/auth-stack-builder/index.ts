export { AmplifyAuthCognitoStack } from './auth-cognito-stack-builder';
export { AmplifyAuthTransform } from './auth-stack-transform';
export { AmplifyUserPoolGroupStack } from './auth-user-pool-group-stack-builder';
export { AmplifyUserPoolGroupTransform } from './user-pool-group-stack-transform';
