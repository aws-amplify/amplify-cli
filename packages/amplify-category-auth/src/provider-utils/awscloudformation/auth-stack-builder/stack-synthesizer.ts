import { ISynthesisSession, Stack, LegacyStackSynthesizer } from '@aws-cdk/core';
import { Template } from 'cloudform-types';
import { AmplifyAuthCognitoStack } from './auth-cognito-stack-builder';
import { AmplifyUserPoolGroupStack, AmplifyUserPoolGroupStackOutputs } from './auth-user-pool-group-stack-builder';

export class AuthStackSythesizer extends LegacyStackSynthesizer {
  private readonly stacks: Map<string, Stack> = new Map();
  private static readonly stackAssets: Map<string, Template> = new Map();

  protected synthesizeStackTemplate(stack: Stack, session: ISynthesisSession): void {
    if (
      stack instanceof AmplifyAuthCognitoStack ||
      stack instanceof AmplifyUserPoolGroupStack ||
      stack instanceof AmplifyUserPoolGroupStackOutputs
    ) {
      this.addStack(stack);
      const template = stack.renderCloudFormationTemplate(session) as string;
      const templateName = stack.node.id;
      this.setStackAsset(templateName, template);
    } else {
      throw new Error(
        'Error synthesizing the template. Expected Stack to be either instance of AmplifyAuthCognitoStack or AmplifyUserPoolGroupStack',
      );
    }
  }

  setStackAsset(templateName: string, template: string): void {
    AuthStackSythesizer.stackAssets.set(templateName, JSON.parse(template));
  }

  collectStacks(): Map<string, Template> {
    return new Map(AuthStackSythesizer.stackAssets.entries());
  }

  addStack(stack: Stack) {
    this.stacks.set(stack.node.id, stack);
  }

  getStack = (stackName: string): Stack => {
    if (this.stacks.has(stackName)) {
      return this.stacks.get(stackName)!;
    }
    throw new Error(`Stack ${stackName} is not created`);
  };
}
