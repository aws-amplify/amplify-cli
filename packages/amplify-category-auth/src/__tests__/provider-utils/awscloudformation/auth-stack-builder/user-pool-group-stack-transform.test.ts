import { AmplifyUserPoolGroupTransform } from '../../../../provider-utils/awscloudformation/auth-stack-builder';
import { $TSContext } from 'amplify-cli-core';

jest.mock('amplify-cli-core', () => ({
  ...(jest.requireActual('amplify-cli-core') as {}),
  stateManager: {
    getLocalEnvInfo: jest.fn().mockReturnValue('testenv'),
  },
  pathManager: {
    getBackendDirPath: jest.fn().mockReturnValue('..'),
  },
  JSONUtilities: {
    readJson: jest
      .fn()
      .mockReturnValueOnce([
        {
          groupName: 'adin',
          precedence: 1,
        },
      ])
      .mockReturnValueOnce([
        {
          groupName: 'adinmock',
          precedence: 2,
        },
      ]),
  },
  buildOverrideDir: jest.fn().mockResolvedValue(false),
  writeCFNTemplate: jest.fn(),
}));

const getCLIInputPayload_mock = jest.fn().mockReturnValue({
  cognitoConfig: {
    identityPoolName: 'extauth387063394_identitypool_87063394',
    resourceName: 'extauth38706339487063394',
  },
});

const isCLIInputsValid_mock = jest.fn().mockReturnValue('true');

jest.mock('../../../../provider-utils/awscloudformation/auth-inputs-manager/auth-input-state.ts', () => {
  return {
    AuthInputState: jest.fn().mockImplementation(() => {
      return {
        getCLIInputPayload: getCLIInputPayload_mock,
        isCLIInputsValid: isCLIInputsValid_mock,
      };
    }),
  };
});

const context_stub = {};

const context_stub_typed = context_stub as unknown as $TSContext;
describe('Check UserPool Group Template', () => {
  it('Generated rootstack template during Push one group', async () => {
    // CFN transform for UserPool Group stack

    const resourceName = 'mockResource';
    const userPoolTransform = new AmplifyUserPoolGroupTransform(resourceName);
    const mock_template = await userPoolTransform.transform(context_stub_typed);
    expect(mock_template).toMatchSnapshot();
  });

  it('Generated rootstack template during Push with two groups', async () => {
    // CFN transform for UserPool Group stack

    const resourceName = 'mockResource';
    const userPoolTransform = new AmplifyUserPoolGroupTransform(resourceName);
    const mock_template = await userPoolTransform.transform(context_stub_typed);
    expect(mock_template).toMatchSnapshot();
  });
});
