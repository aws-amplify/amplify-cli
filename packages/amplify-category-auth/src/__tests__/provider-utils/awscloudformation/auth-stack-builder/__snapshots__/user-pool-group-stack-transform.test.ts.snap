// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check UserPool Group Template Generated rootstack template during Push one group 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": Object {
    "ShouldNotCreateEnvResources": Object {
      "Fn::Equals": Array [
        Object {
          "Ref": "env",
        },
        "NONE",
      ],
    },
  },
  "Description": "Root Stack for AWS Amplify CLI",
  "Outputs": Object {
    "adinGroupRole": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "adinGroupRole",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "AuthRoleArn": Object {
      "Type": "String",
    },
    "UnauthRoleArn": Object {
      "Type": "String",
    },
    "authmockResourceAppClientID": Object {
      "Default": "authmockResourceAppClientID",
      "Type": "String",
    },
    "authmockResourceAppClientIDWeb": Object {
      "Default": "authmockResourceAppClientIDWeb",
      "Type": "String",
    },
    "authmockResourceIdentityPoolId": Object {
      "Default": "authmockResourceIdentityPoolId",
      "Type": "String",
    },
    "authmockResourceUserPoolId": Object {
      "Default": "authmockResourceUserPoolId",
      "Type": "String",
    },
    "env": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "LambdaExecutionRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:logs:*:*:*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "UserGroupLogPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "cognito-identity:SetIdentityPoolRoles",
                    "cognito-identity:ListIdentityPools",
                    "cognito-identity:describeIdentityPool",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "UserGroupExecutionPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:PassRole",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Ref": "AuthRoleArn",
                    },
                    Object {
                      "Ref": "UnauthRoleArn",
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "UserGroupPassRolePolicy",
          },
        ],
        "RoleName": Object {
          "Fn::If": Array [
            "ShouldNotCreateEnvResources",
            "mockResource",
            Object {
              "Fn::Join": Array [
                "",
                Array [
                  "mockResource-ExecutionRole-",
                  Object {
                    "Ref": "env",
                  },
                ],
              ],
            },
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "RoleMapFunction": Object {
      "Properties": Object {
        "Code": Object {
          "ZipFile": "const response = require('cfn-response');
const AWS = require('aws-sdk');
exports.handler = (event, context) => {
  if (event.RequestType == 'Delete') {
    response.send(event, context, response.SUCCESS, { message: 'Request type delete' });
  }
  if (event.RequestType == 'Create' || event.RequestType == 'Update') {
    let { identityPoolId, appClientID, appClientIDWeb, userPoolId, region } = event.ResourceProperties;
    try {
      const cognitoidentity = new AWS.CognitoIdentity();
      let params = {
        IdentityPoolId: identityPoolId,
        Roles: {
          authenticated: event.ResourceProperties.AuthRoleArn,
          unauthenticated: event.ResourceProperties.UnauthRoleArn,
        },
        RoleMappings: {},
      };
      if (appClientIDWeb) {
        params.RoleMappings[\`cognito-idp.\${region}.amazonaws.com/\${userPoolId}:\${appClientIDWeb}\`] = {
          Type: 'Token',
          AmbiguousRoleResolution: 'AuthenticatedRole',
        };
      }
      if (appClientID) {
        params.RoleMappings[\`cognito-idp.\${region}.amazonaws.com/\${userPoolId}:\${appClientID}\`] = {
          Type: 'Token',
          AmbiguousRoleResolution: 'AuthenticatedRole',
        };
      }
      cognitoidentity.setIdentityPoolRoles(params).promise();
      response.send(event, context, response.SUCCESS, { message: 'Successfully updated identity pool.' });
    } catch (err) {
      response.send(event, context, response.FAILED, { message: 'Error updating identity pool' });
    }
  }
};
",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaExecutionRole",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RoleMapFunctionInput": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "RoleMapFunction",
      ],
      "Properties": Object {
        "AuthRoleArn": Object {
          "Ref": "AuthRoleArn",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "RoleMapFunction",
            "Arn",
          ],
        },
        "UnauthRoleArn": Object {
          "Ref": "UnauthRoleArn",
        },
        "appClientID": Object {
          "Ref": "authmockResourceAppClientID",
        },
        "appClientIDWeb": Object {
          "Ref": "authmockResourceAppClientIDWeb",
        },
        "env": Object {
          "Ref": "env",
        },
        "identityPoolId": Object {
          "Ref": "authmockResourceIdentityPoolId",
        },
        "region": Object {
          "Ref": "AWS::Region",
        },
        "userPoolId": Object {
          "Ref": "authmockResourceUserPoolId",
        },
      },
      "Type": "Custom::LambdaCallout",
      "UpdateReplacePolicy": "Delete",
    },
    "adinGroup": Object {
      "Properties": Object {
        "Description": "override success",
        "GroupName": "adin",
        "Precedence": 1,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "adinGroupRole",
            "Arn",
          ],
        },
        "UserPoolId": Object {
          "Ref": "authmockResourceUserPoolId",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "adinGroupRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "authmockResourceIdentityPoolId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
              "Sid": "",
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "authmockResourceUserPoolId",
              },
              "-adinGroupRole",
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;

exports[`Check UserPool Group Template Generated rootstack template during Push with two groups 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Conditions": Object {
    "ShouldNotCreateEnvResources": Object {
      "Fn::Equals": Array [
        Object {
          "Ref": "env",
        },
        "NONE",
      ],
    },
  },
  "Description": "Root Stack for AWS Amplify CLI",
  "Outputs": Object {
    "adinmockGroupRole": Object {
      "Value": Object {
        "Fn::GetAtt": Array [
          "adinmockGroupRole",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "AuthRoleArn": Object {
      "Type": "String",
    },
    "UnauthRoleArn": Object {
      "Type": "String",
    },
    "authmockResourceAppClientID": Object {
      "Default": "authmockResourceAppClientID",
      "Type": "String",
    },
    "authmockResourceAppClientIDWeb": Object {
      "Default": "authmockResourceAppClientIDWeb",
      "Type": "String",
    },
    "authmockResourceIdentityPoolId": Object {
      "Default": "authmockResourceIdentityPoolId",
      "Type": "String",
    },
    "authmockResourceUserPoolId": Object {
      "Default": "authmockResourceUserPoolId",
      "Type": "String",
    },
    "env": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "LambdaExecutionRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "sts:AssumeRole",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "Service": Array [
                  "lambda.amazonaws.com",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": "arn:aws:logs:*:*:*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "UserGroupLogPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "cognito-identity:SetIdentityPoolRoles",
                    "cognito-identity:ListIdentityPools",
                    "cognito-identity:describeIdentityPool",
                  ],
                  "Effect": "Allow",
                  "Resource": "*",
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "UserGroupExecutionPolicy",
          },
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "iam:PassRole",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    Object {
                      "Ref": "AuthRoleArn",
                    },
                    Object {
                      "Ref": "UnauthRoleArn",
                    },
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "UserGroupPassRolePolicy",
          },
        ],
        "RoleName": Object {
          "Fn::If": Array [
            "ShouldNotCreateEnvResources",
            "mockResource",
            Object {
              "Fn::Join": Array [
                "",
                Array [
                  "mockResource-ExecutionRole-",
                  Object {
                    "Ref": "env",
                  },
                ],
              ],
            },
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "RoleMapFunction": Object {
      "Properties": Object {
        "Code": Object {
          "ZipFile": "const response = require('cfn-response');
const AWS = require('aws-sdk');
exports.handler = (event, context) => {
  if (event.RequestType == 'Delete') {
    response.send(event, context, response.SUCCESS, { message: 'Request type delete' });
  }
  if (event.RequestType == 'Create' || event.RequestType == 'Update') {
    let { identityPoolId, appClientID, appClientIDWeb, userPoolId, region } = event.ResourceProperties;
    try {
      const cognitoidentity = new AWS.CognitoIdentity();
      let params = {
        IdentityPoolId: identityPoolId,
        Roles: {
          authenticated: event.ResourceProperties.AuthRoleArn,
          unauthenticated: event.ResourceProperties.UnauthRoleArn,
        },
        RoleMappings: {},
      };
      if (appClientIDWeb) {
        params.RoleMappings[\`cognito-idp.\${region}.amazonaws.com/\${userPoolId}:\${appClientIDWeb}\`] = {
          Type: 'Token',
          AmbiguousRoleResolution: 'AuthenticatedRole',
        };
      }
      if (appClientID) {
        params.RoleMappings[\`cognito-idp.\${region}.amazonaws.com/\${userPoolId}:\${appClientID}\`] = {
          Type: 'Token',
          AmbiguousRoleResolution: 'AuthenticatedRole',
        };
      }
      cognitoidentity.setIdentityPoolRoles(params).promise();
      response.send(event, context, response.SUCCESS, { message: 'Successfully updated identity pool.' });
    } catch (err) {
      response.send(event, context, response.FAILED, { message: 'Error updating identity pool' });
    }
  }
};
",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LambdaExecutionRole",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "RoleMapFunctionInput": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "RoleMapFunction",
      ],
      "Properties": Object {
        "AuthRoleArn": Object {
          "Ref": "AuthRoleArn",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "RoleMapFunction",
            "Arn",
          ],
        },
        "UnauthRoleArn": Object {
          "Ref": "UnauthRoleArn",
        },
        "appClientID": Object {
          "Ref": "authmockResourceAppClientID",
        },
        "appClientIDWeb": Object {
          "Ref": "authmockResourceAppClientIDWeb",
        },
        "env": Object {
          "Ref": "env",
        },
        "identityPoolId": Object {
          "Ref": "authmockResourceIdentityPoolId",
        },
        "region": Object {
          "Ref": "AWS::Region",
        },
        "userPoolId": Object {
          "Ref": "authmockResourceUserPoolId",
        },
      },
      "Type": "Custom::LambdaCallout",
      "UpdateReplacePolicy": "Delete",
    },
    "adinmockGroup": Object {
      "Properties": Object {
        "Description": "override success",
        "GroupName": "adinmock",
        "Precedence": 2,
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "adinmockGroupRole",
            "Arn",
          ],
        },
        "UserPoolId": Object {
          "Ref": "authmockResourceUserPoolId",
        },
      },
      "Type": "AWS::Cognito::UserPoolGroup",
    },
    "adinmockGroupRole": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": Object {
                "ForAnyValue:StringLike": Object {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": Object {
                  "cognito-identity.amazonaws.com:aud": Object {
                    "Ref": "authmockResourceIdentityPoolId",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": Object {
                "Federated": "cognito-identity.amazonaws.com",
              },
              "Sid": "",
            },
          ],
          "Version": "2012-10-17",
        },
        "RoleName": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "authmockResourceUserPoolId",
              },
              "-adinmockGroupRole",
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
